keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,462 @@
+
+#### Enhancements in CSI Node Plugins
+
+Code for CSI Node Plugins need to be refactored to support CSI Node APIs in both Linux and Windows nodes. While the code targeting Linux nodes can assume privileged access to the host, the code targeting Windows nodes need to invoke the GRPC client API associated with the desired version of the CSIProxyService described above. CSI Node Plugins that will use the Kubernetes/Utils/mounter package introduced in this [PR](https://github.com/kubernetes/utils/pull/100/files) will require minimal platform specific code targeting Windows and Linux.
+
+Once compiled for Windows, container images based on Window Base images (like NanoServer) needs to be published and maintained. Container images targeting Linux nodes will need to be based on the desired Linux distro base image.
+
+New YAMLs for DaemonSets associated with CSI Node Plugins needs to be authored that will (1) target Windows nodes and (2) use Windows paths for domain socket related paths as illustrated in the User Story section above.
+
+### Risks and Mitigations
+
+Any pod on a Windows node can be configured to mount `\\.\pipe\csi-proxy-v[N]` and perform privileged operations. Thus csi-proxy presents a potential security risk. To mitigate the risk, here are some options:
+
+#### Mitigation using PSP
+Administrators can enable Pod Security Policy (PSP) in their cluster and configure PSPs to:
+1. Disallow host path mounts as part of a default cluster-wide PSP. This will affect all pods in the cluster across Linux and Windows that mount any host paths.
+2. Allow host path mounts with pathPrefix = `\\.\pipe\csi-proxy`. Restrict usage of this PSP to only SAs associated with the DaemonSets of CSI Node Plugins.
+Support will need to be implemented in AllowsHostVolumePath to handle Windows pipe paths.
+
+#### Mitigation using a webhook
+An admission webhook can be implemented and deployed in clusters with Windows nodes that will reject all containers that mount `\\.\pipe\csi-proxy` as a hostPath volume but does not have privileged flag set in the pod's securityContext specification. This allows the privileged setting to be used for Windows pods as an indication the container will perform privileged operations. Other cluster-wide policies (e.g. PSP) that act on the privileged setting in a container's securityContext can enforce the same for CSI Node plugins targeting Windows nodes. Note that this does not in any way change how the privileged setting is used today for Linux nodes. If in the future, full privileged container support is introduced in Windows (as in Linux today), functionality of existing CSI Node Plugin DaemonSets (targeting Windows) with the privileged flag set should not get negatively impacted as they will be launched as privileged containers.
```

Comment:
I'm not sure how symlinks are handled on windows, but on linux this would be easy to circumvent.

In general, mounting of hostpaths should be thought of as a highly privileged operation, especially if they aren't read-only (can these be mounted read-only?). If the only way to get access to the socket is via a hostpath mount, I'm not too worried.

If we did want to add more enforcement, I'd prefer to add authentication & authorization to the socket itself. E.g. ues token review + subject access review to ensure the container using the proxy is authorized to do so. I'm not sure this is necessary though.

Reply:
symlinks to named pipes are not supported in Windows.